gazenot/src/client.rs

```diff
@@ -0,0 +1,672 @@
+        let url = Url::from_str(&format!("{base}/{filename}"))?;
+        Ok(url)
+    }
+
+    pub fn upload_artifact_set_url(&self, set: &ArtifactSet, filename: &str) -> ResultInner<Url> {
+        // POST /:sourcehost/:owner/:package/artifacts/:id/
+        let base = set.upload_url.clone().unwrap_or_else(|| {
+            let server = &self.api_server;
+            let source_host = &self.source_host;
+            let owner = &self.owner;
+            let ArtifactSet {
+                package, public_id, ..
+            } = set;
+            format!("https://{server}/{source_host}/{owner}/{package}/artifacts/{public_id}/upload")
+        });
+        let url = Url::from_str(&format!("{base}/{filename}"))?;
+        Ok(url)
+    }
+
+    pub fn create_release_url(&self, set: &ArtifactSet) -> ResultInner<Url> {
+        // POST /:sourcehost/:owner/:package/releases
+        let url = set.release_url.clone().unwrap_or_else(|| {
+            let server = &self.api_server;
+            let source_host = &self.source_host;
+            let owner = &self.owner;
+            let package = &set.package;
+            format!("https://{server}/{source_host}/{owner}/{package}/releases")
+        });
+        let url = Url::from_str(&url)?;
+        Ok(url)
+    }
+
+    pub fn create_announcement_url(&self, release: &Release) -> ResultInner<Url> {
+        // POST /:sourcehost/:owner/announcements
+        let url = release.announce_url.clone().unwrap_or_else(|| {
+            let server = &self.api_server;
+            let source_host = &self.source_host;
+            let owner = &self.owner;
+            format!("https://{server}/{source_host}/{owner}/announcements")
+        });
+        let url = Url::from_str(&url)?;
+        Ok(url)
+    }
+
+    pub fn list_releases_url(&self, package: &PackageName) -> ResultInner<Url> {
+        // GET /:sourcehost/:owner/:projects/releases
+        let server = &self.api_server;
+        let source_host = &self.source_host;
+        let owner = &self.owner;
+        let package = &package;
+        let url = Url::from_str(&format!(
+            "https://{server}/{source_host}{owner}/{package}/releases"
+        ))?;
+        Ok(url)
+    }
+}
+
+async fn join_all<T>(
+    queries: impl IntoIterator<Item = (String, Url, tokio::task::JoinHandle<ResultInner<T>>)>,
+) -> Result<Vec<T>> {
+    let mut results = Vec::new();
+    for (desc, url, query) in queries {
+        let result = query
+            .await
+            .map_err(|e| GazenotError::with_url(&desc, url.to_string(), e))?
+            .map_err(|e| GazenotError::with_url(&desc, url.to_string(), e))?;
+        results.push(result);
+    }
+    Ok(results)
+}
+
+fn auth_headers(source: &SourceHost, owner: &Owner) -> ResultInner<HeaderMap> {
+    // extra-awkard code so you're on your toes and properly treat this like radioactive waste
+    // DO NOT UNDER ANY CIRCUMSTANCES PRINT THIS VALUE.
+    // DO NOT IMPLEMENT DEBUG ON Abyss OR AbyssInner!!
```

Comment:
Can you add versions of these comments on the `Abyss` and `AbyssInner` definitions? It'll be important to have it flagged up there as well to help avoid it happening by accident.